src/graph/caching.h

```diff
@@ -0,0 +1,52 @@
+/*
+   IGraph library.
+   Copyright (C) 2022  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef IGRAPH_CACHING_H
+#define IGRAPH_CACHING_H
+
+#include "igraph_datatype.h"
+#include "igraph_decls.h"
+#include "igraph_error.h"
+#include "igraph_types.h"
+
+#include "internal/hacks.h"
+
+#include <string.h> /* memset */
+
+__BEGIN_DECLS
+
+struct igraph_i_property_cache_t {
+    igraph_bool_t value[IGRAPH_PROP_I_SIZE];
+
+    /** Bit field that stores which of the properties are cached at the moment */
+    uint32_t known;
+};
+
+igraph_error_t igraph_i_property_cache_init(igraph_i_property_cache_t *cache);
+igraph_error_t igraph_i_property_cache_copy(
+        igraph_i_property_cache_t *cache,
+        const igraph_i_property_cache_t *other_cache);
+void igraph_i_property_cache_destroy(igraph_i_property_cache_t *cache);
```

Comment:
As I remember, `init`/`copy`/`destroy` was my design originally, but looking at it again, it seems that they are never used without `calloc` and `free`. Should we have included memory allocation / freeing directly in these functions?

Anyway, refactoring this wouldn't bring any gains, so let's leave as is.

Reply:
Yes, I think that may be the reason why I did it this way.